File: src/board.js
const { key, neighbors, directionIndex, directions, add } = require('./coords');

const cloneBoard = (board) => {
  const next = new Map();
  for (const [k, stack] of board.entries()) {
    next.set(k, [...stack]);
  }
  return next;
};

const getStack = (board, coord) => board.get(key(coord)) || [];

const isOccupied = (board, coord) => getStack(board, coord).length > 0;

const topPiece = (board, coord) => {
  const stack = getStack(board, coord);
  if (!stack.length) return null;
  return stack[stack.length - 1];
};

const placePiece = (board, coord, piece) => {
  const k = key(coord);
  const existing = board.get(k) || [];
  board.set(k, [...existing, piece]);
};

const removeTopPiece = (board, coord) => {
  const k = key(coord);
  const existing = board.get(k) || [];
  const next = [...existing];
  const removed = next.pop();
  if (next.length === 0) {
    board.delete(k);
  } else {
    board.set(k, next);
  }
  return removed;
};

const moveTopPiece = (board, from, to) => {
  const piece = removeTopPiece(board, from);
  if (!piece) return null;
  placePiece(board, to, piece);
  return piece;
};

const occupiedCoords = (board) =>
  Array.from(board.entries())
    .filter(([, stack]) => stack.length)
    .map(([k]) => k);

const adjacencyGraphFromBoard = (board) => {
  const occ = new Set(occupiedCoords(board));
  const graph = new Map();
  for (const k of occ) {
    const coord = k.split(',').map(Number);
    const nodeCoord = { x: coord[0], y: coord[1], z: coord[2] };
    const adj = neighbors(nodeCoord)
      .map((n) => key(n))
      .filter((nk) => occ.has(nk));
    graph.set(k, adj);
  }
  return graph;
};

const isHiveConnected = (board) => {
  const occKeys = occupiedCoords(board);
  if (occKeys.length <= 1) return true;
  const visited = new Set();
  const queue = [occKeys[0]];
  visited.add(occKeys[0]);
  while (queue.length) {
    const current = queue.shift();
    const [x, y, z] = current.split(',').map(Number);
    for (const n of neighbors({ x, y, z })) {
      const nk = key(n);
      if (visited.has(nk)) continue;
      if (board.has(nk)) {
        visited.add(nk);
        queue.push(nk);
      }
    }
  }
  return visited.size === occKeys.length;
};

const wouldBreakHive = (board, coord) => {
  const k = key(coord);
  const stack = board.get(k);
  if (!stack || stack.length === 0) return false;
  const temp = cloneBoard(board);
  removeTopPiece(temp, coord);
  return !isHiveConnected(temp);
};

const canSlideBetween = (board, from, to) => {
  const dirIdx = directionIndex(from, to);
  if (dirIdx === -1) return false;
  const leftIdx = (dirIdx + 5) % 6;
  const rightIdx = (dirIdx + 1) % 6;
  const leftNeighbor = add(from, directions[leftIdx]);
  const rightNeighbor = add(from, directions[rightIdx]);
  const leftOccupied = isOccupied(board, leftNeighbor);
  const rightOccupied = isOccupied(board, rightNeighbor);
  return !(leftOccupied && rightOccupied);
};

module.exports = {
  cloneBoard,
  getStack,
  isOccupied,
  topPiece,
  placePiece,
  removeTopPiece,
  moveTopPiece,
  occupiedCoords,
  adjacencyGraphFromBoard,
  isHiveConnected,
  wouldBreakHive,
  canSlideBetween,
};

File: src/board.py
from __future__ import annotations

from copy import deepcopy
from typing import Dict, List, Optional

from .coords import Coord, DIRECTIONS, add, neighbors, direction_index

Piece = Dict[str, str]
Board = Dict[str, List[Piece]]


def clone_board(board: Board) -> Board:
    return deepcopy(board)


def get_stack(board: Board, coord: Coord) -> List[Piece]:
    return board.get(coord.key(), [])


def is_occupied(board: Board, coord: Coord) -> bool:
    return len(get_stack(board, coord)) > 0


def top_piece(board: Board, coord: Coord) -> Optional[Piece]:
    stack = get_stack(board, coord)
    return stack[-1] if stack else None


def place_piece(board: Board, coord: Coord, piece: Piece) -> None:
    board.setdefault(coord.key(), []).append(piece)


def remove_top_piece(board: Board, coord: Coord) -> Optional[Piece]:
    stack = board.get(coord.key())
    if not stack:
        return None
    piece = stack.pop()
    if not stack:
        board.pop(coord.key(), None)
    return piece


def move_top_piece(board: Board, from_coord: Coord, to_coord: Coord) -> Optional[Piece]:
    piece = remove_top_piece(board, from_coord)
    if piece is None:
        return None
    place_piece(board, to_coord, piece)
    return piece


def occupied_coords(board: Board) -> List[str]:
    return [k for k, stack in board.items() if stack]


def is_hive_connected(board: Board) -> bool:
    occ_keys = occupied_coords(board)
    if len(occ_keys) <= 1:
        return True
    visited = set()
    queue = [occ_keys[0]]
    visited.add(occ_keys[0])
    while queue:
        current_key = queue.pop(0)
        cx, cy, cz = map(int, current_key.split(","))
        for n in neighbors(Coord(cx, cy, cz)):
            nk = n.key()
            if nk in visited:
                continue
            if nk in board:
                visited.add(nk)
                queue.append(nk)
    return len(visited) == len(occ_keys)


def would_break_hive(board: Board, coord: Coord) -> bool:
    if coord.key() not in board:
        return False
    temp = clone_board(board)
    remove_top_piece(temp, coord)
    return not is_hive_connected(temp)


def can_slide_between(board: Board, from_coord: Coord, to_coord: Coord) -> bool:
    dir_idx = direction_index(from_coord, to_coord)
    if dir_idx == -1:
        return False
    left_idx = (dir_idx + 5) % 6
    right_idx = (dir_idx + 1) % 6
    left_neighbor = add(from_coord, DIRECTIONS[left_idx])
    right_neighbor = add(from_coord, DIRECTIONS[right_idx])
    left_occupied = is_occupied(board, left_neighbor)
    right_occupied = is_occupied(board, right_neighbor)
    return not (left_occupied and right_occupied)
